external/llvm-project/mlir/lib/Dialect/Tosa/Transforms/TosaPartition.cpp

```diff
@@ -321,21 +322,66 @@ void outlineConvPartOps(Operation *convOp, ArrayRef<Operation *> secondOps,
 
   // Erase the ops we outlined, which should be safe now.
   for (auto &op : llvm::make_early_inc_range(llvm::reverse(secondOps))) {
-    assert(op->use_empty() && "expected 'op' to have no uses");
-    op->erase();
+    if (op->use_empty()) {
+      op->erase();
+    }
   }
   assert(convOp->use_empty() && "expected 'op' to have no uses");
   convOp->erase();
   for (auto &op : llvm::make_early_inc_range(frontOps)) {
-    assert(op->use_empty() && "expected 'op' to have no uses");
-    op->erase();
+    if (op->use_empty()) {
+      op->erase();
+    }
   }
 }
 
 // Inspired by / adapted from TestSCFIfUtilsPass in
 // test/lib/Transforms/TestSCFUtils.cpp.
 class TosaPartitionPass : public TosaPartitionBase<TosaPartitionPass> {
 public:
+  bool isConstantZero(Operation *op) {
+    // test for zero
+    bool is_zero = false;
+    if (auto cst = dyn_cast<arith::ConstantOp>(op)) {
+      if (auto val = cst.getValue().dyn_cast<ElementsAttr>()) {
+        auto valType = val.getType().dyn_cast<ShapedType>();
+        auto valElemType = valType.getElementType();
+        if (valElemType.isa<FloatType>()) {
+          is_zero = true;
+          for (auto ii = val.value_begin<APFloat>();
+               is_zero && ii != val.value_end<APFloat>(); ++ii)
+            is_zero &= (*ii).isZero();
+        } else if (valElemType.isa<IntegerType>()) {
+          is_zero = true;
+          for (auto ii = val.value_begin<APInt>();
+               is_zero && ii != val.value_end<APInt>(); ++ii)
+            is_zero &= (*ii).isZero();
+        }
+      } else if (auto val = cst.getValue().dyn_cast<FloatAttr>()) {
+        return val.getValue().isZero();
+      } else if (auto val = cst.getValue().dyn_cast<IntegerAttr>()) {
+        return val.getValue().isZero();
+      }
+    }
+    return is_zero;
+  }
+
+  void traceInputs(Operation *op, SmallVector<Operation *> &predecessors,
+                   SetVector<Value> &inputNodes) {
+    for (const auto &opnd : op->getOperands()) {
+      Operation *usedOp = opnd.getDefiningOp();
+      if (usedOp && (isConstantZero(usedOp) || isElementwiseOp(usedOp))) {
```

Comment:
I had a hasOneUse() test in here because we're absorbing the op into the outlined kernel, removing the original, and there's no way for a second use to see it after that.  I notice that you just don't erase nodes that still have uses, but is that sufficient for correct behavior?

Reply:
Yes, it is sufficient because we clone the op, it is not actually moved. Then later if all uses of the original are deleted it gets removed.